### PR TITLE
Document repr packed(N).

### LINKED ISSUE #483: Document `#[repr(packed(N))]`
- Version: 1.33
- PR: https://github.com/rust-lang/rust/pull/57049
- Tracking issue: https://github.com/rust-lang/rust/issues/33158
- RFC: https://github.com/rust-lang/rfcs/pull/1399

The implementation diverged from the RFC a bit...

cc @cramertj

### PR DESCRIPTION
**Pull request by ehuss**
Closes #483

I feel like this is lacking, but I'm not sure what else to add.

### CODE REVIEW

File: src/type-layout.md [hunk 1]
```diff
@@ -280,19 +280,29 @@ The `align` representation can be used on `struct`s and `union`s to raise the
...
+alignment and padding of the type.
+
+The packing value is specified as a parameter in the form of
+`#[repr(packed(x))]`. If no value is given, as in `#[repr(packed)]`, then the
+packing value is 1. The packing value must be a power of two up to
+2<sup>29</sup>.
 
-It modifies the representation (either the default or `C`) by removing any
-padding bytes and forcing the alignment of the type to `1`.
+The `packed` representation sets the alignment to the smaller of the specified
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
The `packed` representation sets the type's alignment to the smaller of the specified
```

File: src/type-layout.md [hunk 2]
```diff
@@ -280,19 +280,29 @@ The `align` representation can be used on `struct`s and `union`s to raise the
...
+packing and the current alignment (either default or `C`). The alignments of
+each field for the purpose of positioning fields would also be the smaller of
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
each field, for the purpose of positioning fields, is the smaller of
```

File: src/type-layout.md [hunk 3]
```diff
@@ -280,19 +280,29 @@ The `align` representation can be used on `struct`s and `union`s to raise the
...
+the specified packing and the alignment of the type of that field. If the
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
the specified packing and the alignment of the field's type. If the
```

File: src/type-layout.md [hunk 4]
```diff
@@ -280,19 +280,29 @@ The `align` representation can be used on `struct`s and `union`s to raise the
...
+
+The packing value is specified as a parameter in the form of
+`#[repr(packed(x))]`. If no value is given, as in `#[repr(packed)]`, then the
+packing value is 1. The packing value must be a power of two up to
+2<sup>29</sup>.
 
-It modifies the representation (either the default or `C`) by removing any
-padding bytes and forcing the alignment of the type to `1`.
+The `packed` representation sets the alignment to the smaller of the specified
+packing and the current alignment (either default or `C`). The alignments of
```
Comment (Centril) [inline src/type-layout.md]: Here the "default" alignment is referenced but above "primitive alignment" is used... we should pick one phrasing.

Comment (Centril) [inline src/type-layout.md]: "or `C`" is not super clear wrt. referring to C-the-language

Comment (Havvy) [reply src/type-layout.md]: Get rid of both "current' and "primitive". Get rid of the parenthetical. Instead spell it out explicitly that if the type has the default or C representation, and the alignment from that representation is smaller, it will use that alignment.

**Comment by ehuss** [comment]
I added an experimental commit to propose a different way of documenting these `repr` instructions. I think it more accurately captures how it works. That is, there are some [basic "representations"](https://github.com/rust-lang/rust/blob/4c27fb19ba15a2e45485e601a79914c6280196b0/src/librustc/ty/mod.rs#L2024-L2035), and `packed` and `align` just alter them. In the relevant cases (on stable) there are only two — `C` and default. I think it makes it clearer about what can be "combined".

I do have some concerns and doubts:
- This introduces new terminology ("modifier") that I made up.
- The descriptions between `align` and `packed` have a lot of duplication.
- I think that `align` and `packed` *force* the alignment to the given value. However, I'm not certain. The [parsing code](https://github.com/rust-lang/rust/blob/4c27fb19ba15a2e45485e601a79914c6280196b0/src/libsyntax/attr/builtin.rs#L732) calls them "hints" in the error messages, but that terminology is not used elsewhere. I'm fairly sure that align is forced [here](https://github.com/rust-lang/rust/blob/4c27fb19ba15a2e45485e601a79914c6280196b0/src/librustc/ty/layout.rs#L347) and packed is forced [here](https://github.com/rust-lang/rust/blob/4c27fb19ba15a2e45485e601a79914c6280196b0/src/librustc/ty/layout.rs#L331).  This 1000 line function is a bit hard to follow.
    - Related: I think that `packed` has a floor of the i8 alignment [here](https://github.com/rust-lang/rust/blob/4c27fb19ba15a2e45485e601a79914c6280196b0/src/librustc/ty/layout.rs#L260). This is always 8-bits on all platforms rustc builds on that I can find. I tried building with a custom target where it was 16-bits, but libcore fails to build with `i8:16:32` alignment (and I need the `align_of` function to test my hypothesis). I'm not sure if this floor should be documented, or if Rust should always assume this to be "1"

File: src/type-layout.md [hunk 5]
```diff
@@ -128,14 +127,36 @@ struct ThreeInts {
 }
 ```
 
+The alignment may be raised or lowered with the `align` and `packed` modifiers
+respectively. They alter the representation specified in the attribute:
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
respectively. They alter the representation specified in the attribute.
If no representation is specified, the default one is altered.
```

File: src/type-layout.md [hunk 6]
```diff
@@ -113,13 +113,12 @@ All user-defined composite types (`struct`s, `enum`s, and `union`s) have a
 *representation* that specifies what the layout is for the type.
 
 The possible representations for a type are the default representation, `C`,
-the primitive representations, `packed`, and `transparent`. Multiple
-representations can be applied to a single type.
+the primitive representations, and `transparent`.
```
Comment (Centril) [inline src/type-layout.md]: Itemize ^

Also make it clear what the "primitive representations" entail (by itemizing all the primitive reprs)

File: src/type-layout.md [hunk 7]
```diff
@@ -128,14 +127,36 @@ struct ThreeInts {
...
 > the same name have the same representation. For example, `Foo<Bar>` and
 > `Foo<Baz>` both have the same representation.
 
-The representation of a type does not change the layout of its fields. For
-example, a struct with a `C` representation that contains a struct `Inner` with
-the default representation will not change the layout of Inner.
+The representation of a type can change the padding between fields, but does
+not change the layout of the fields themselves. For example, a struct with a
+`C` representation that contains a struct `Inner` with the default
+representation will not change the layout of Inner.
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
representation will not change the layout of `Inner`.
```

File: src/type-layout.md [hunk 8]
```diff
@@ -274,28 +295,42 @@ For all other enumerations, the layout is unspecified.
...
 
-Alignment is specified as a parameter in the form of `#[repr(align(x))]`. The
-alignment value must be a power of two of type `u32`. The `align` representation
-can raise the alignment of a type to be greater than it's primitive alignment,
-it cannot lower the alignment of a type.
+The alignment is specified as an integer parameter in the form of
+`#[repr(align(x))]`. The alignment value must be a power of two from 1 up to
+2<sup>29</sup>.
+
+The `align` modifier raises the type's alignment to the specified alignment.
```
Comment (Centril) [inline src/type-layout.md]: This is a partial repetition of what you said in the leading paragraph.

File: src/type-layout.md [hunk 9]
```diff
@@ -274,28 +295,42 @@ For all other enumerations, the layout is unspecified.
...
+If the specified alignment is less than the alignment of the type without the
```
Comment (Centril) [inline src/type-layout.md]: I believe this concept of "without the ** modifier" is referred to as the "natural **", for example "natural alignment". I think it would be good to introduce such terms earlier (when we first start to speak about alignment) and then use them.

File: src/type-layout.md [hunk 10]
```diff
@@ -274,28 +295,42 @@ For all other enumerations, the layout is unspecified.
...
+### The `packed` modifier
+
+The `packed` modifier can be used on `struct`s and `union`s to lower the
+alignment of the type to a given value.
 
-The `packed` representation can only be used on `struct`s and `union`s.
+The packing value is specified as an integer parameter in the form of
+`#[repr(packed(x))]`. If no value is given, as in `#[repr(packed)]`, then the
+packing value is 1. The packing value must be a power of two from 1 up to
+2<sup>29</sup>.
```
Comment (Centril) [inline src/type-layout.md]: If you extract out the wording about the default it is the same as the paragraph for `align`. So extract these two out into a common notion `repr_value` or something and then reference that one.

File: src/type-layout.md [hunk 11]
```diff
@@ -274,28 +295,42 @@ For all other enumerations, the layout is unspecified.
...
 
-It modifies the representation (either the default or `C`) by removing any
-padding bytes and forcing the alignment of the type to `1`.
+The `packed` modifier lowers the type's alignment to the specified packing. If
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
The `packed` modifier lowers the type's alignment to the specified value. If
```

File: src/type-layout.md [hunk 12]
```diff
@@ -274,28 +295,42 @@ For all other enumerations, the layout is unspecified.
...
+the specified packing is greater to the alignment of the type without the
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
the specified value is greater than the type's alignment without the
```

File: src/type-layout.md [hunk 13]
```diff
@@ -274,28 +295,42 @@ For all other enumerations, the layout is unspecified.
...
+only be applied to the default and `C` representations.
 
-### The `packed` Representation
+### The `packed` modifier
+
+The `packed` modifier can be used on `struct`s and `union`s to lower the
+alignment of the type to a given value.
 
-The `packed` representation can only be used on `struct`s and `union`s.
+The packing value is specified as an integer parameter in the form of
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
The packing value (hence: "value") is specified as an integer parameter in the form of
```

File: src/type-layout.md [hunk 14]
```diff
@@ -274,28 +295,42 @@ For all other enumerations, the layout is unspecified.
...
+`#[repr(packed(x))]`. If no value is given, as in `#[repr(packed)]`, then the
+packing value is 1. The packing value must be a power of two from 1 up to
```
Comment (Centril) [inline src/type-layout.md]:
```suggestion
value is 1. The value must be a power of two from 1 up to
```

**Comment by ehuss** [comment]
I attempted to reduce some of the duplication.  It still seems a little awkward, though.

**Comment by Centril** [comment]
I think this is strict improvement for now. Thanks @ehuss.
Feel free to merge at will :)

**Comment by retep998** [comment]
Technically it would be more accurate to say `packed` lowers the alignment of the *fields* in the struct, rather than the struct itself. But since the alignment of the struct itself is determined from the maximum alignment of the fields, the struct alignment effectively decreases as well. `align` meanwhile increases the alignment of the struct but *not* its fields.

**Comment by Centril** [comment]
@retep998 Could you file a clarification PR?
